### ISSUE TITLE
Invalid DateTime causing DateTime parse exception when using ConvertZeroDatetime=true

### ISSUE THREAD

**Issue by badbod99** [mysql-net/MySqlConnector#527]
![errornulldate](https://user-images.githubusercontent.com/7858558/42375072-4e508f9a-8112-11e8-8c33-d20d7a411e6b.jpg)

This is a stored procedure which returns a column containing 1 DateTime column.  The column contains some null values.  If I use IFNULL(datefield, '') in my stored procedure, no error occurs.  If not, I get the attached error.

I can't see in the code of GetValues or GetValue where a value of DBNull.Value is handled.

**Comment by bgrainger** [comment]
`Row.GetValue` handles a `NULL` value here: https://github.com/mysql-net/MySqlConnector/blob/29fbed7cd7dbef1e9dbe308fe483e9e6fb47309e/src/MySqlConnector/Core/Row.cs#L347-L348

Using `IFNULL(datefield, '')` will coerce the `DateTime` column to a string, which will disable DateTime parsing for all rows; this is probably why you don't get the exception.

Based on the exception call stack, it looks like the value is formatted with the right pattern, e.g., `yyyy-mm-dd`, but the values aren't valid, e.g., `2018-02-31`.

* Do you have [`ALLOW_INVALID_DATES`](https://dev.mysql.com/doc/refman/8.0/en/sql-mode.html#sqlmode_allow_invalid_dates) enabled on your server?
* Are there invalid "dates" in your data?
* Are you using `AllowZeroDateTime=true` in your connection string?

**Comment by bgrainger** [comment]
If you are using `AllowZeroDateTime=true`, there was a bug that prevented invalid dates (permitted by `ALLOW_INVALID_DATES`) to be retrieved: #529 

That's now fixed in [0.42.3](https://github.com/mysql-net/MySqlConnector/releases/tag/0.42.3), which may fix your exception.

**Comment by badbod99** [comment]
Possibly related but I am not using AllowZeroDateTime but ConvertZeroDateTime instead.

The DateTime values are null in the database, if I convert these to string, i.e. IFNULL(datevalue, ''), no exception is thrown.  Other DateTime values (non-null) are correctly returned in this instance.

* Do you have ALLOW_INVALID_DATES enabled on your server?
>> Yes

* Are there invalid "dates" in your data?
>> No

* Are you using AllowZeroDateTime=true in your connection string?
>> No, using ConvertDateTime=true

I'll try to put together a simple project to reproduce, but the scenario is a DateTime column with valid values and some null.

**Comment by badbod99** [comment]
Ok, I've tried to create a simple project to repro.  Seems the issue is related to invalid dates.  Confusingly, the server is set to not allow invalid dates and the editors shows those dates as null, but I think it's because the database was upgraded from an old version.  Trying to insert a new invalid date fails.

The nulls are being handled correctly by MySqlConnector, it's just MySql is treating invalid dates as null for IFNULL in my case.

So basically, this is not a bug.  But the nicest behaviour would be to convert any invalid date to 1900-01-01 if using ConvertZeroDateTime.

**Comment by bgrainger** [comment]
> But the nicest behaviour would be to convert any invalid date to 1900-01-01 if using ConvertZeroDateTime.

`ConvertZeroDateTime=True` should convert any invalid date to `DateTime.MinValue`, i.e., `0001-01-01 00:00:00`. Are you saying there are cases where this doesn't happen?

Note that `ConvertZeroDateTime=True` does not affect the handling of `NULL` values; `NULL` is considered a valid value in a `DATE` column that allows NULLs.

**Comment by badbod99** [comment]
Confirmed that ConvertZeroDateTime does not convert invalid dates correctly, only zero dates.

SQL

```sql
SET SQL_MODE='TRADITIONAL,ALLOW_INVALID_DATES';

INSERT INTO possnulltest (possnulldate)
VALUES ('0115-02-30 00:00:00');

CREATE TABLE `possnulltest` (
  `pk` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `possnulldate` datetime NULL,
  PRIMARY KEY (`pk`)
) ENGINE=InnoDB AUTO_INCREMENT=1 DEFAULT CHARSET=utf8 ROW_FORMAT=DYNAMIC;
```

C#

```C#
static void Main(string[] args)
{
    using (MySqlConnection conn = new MySqlConnection("datasource=localhost;username=******;password=******;database=plus;Pooling=true;Max Pool Size=80;Min Pool Size=50;default command timeout=60;Convert Zero Datetime=true;"))
    {
        conn.Open();
        using (MySqlCommand comm = conn.CreateCommand())
        {
            comm.CommandText = "SELECT possnulldate FROM possnulltest;";
            comm.CommandType = CommandType.Text;

            using (MySqlDataReader reader = comm.ExecuteReader())
            {
                while (reader.Read())
                {
                    try
                    {
                        if (reader["possnulldate"] == DBNull.Value)
                        {
                            Console.WriteLine("DBNull");
                        }
                        else
                        {
                            var val = reader["possnulldate"];
                            Console.WriteLine(string.Format("{0}", val));
                        }
                    }
                    catch (System.FormatException ex)
                    {
                        Console.WriteLine(ex.ToString());
                    }
                }
            }
        }
    }

    Console.WriteLine("End");
    Console.ReadLine();
}
```

Exception raised:

```
System.FormatException: Couldn't interpret '0115-02-30 00:00:00' as a valid DateTime ---> System.ArgumentOutOfRangeException: Year, Month, and Day parameters describe an un-representable DateTime.
   at System.DateTime.DateToTicks(Int32 year, Int32 month, Int32 day)
   at System.DateTime..ctor(Int32 year, Int32 month, Int32 day, Int32 hour, Int32 minute, Int32 second, Int32 millisecond, DateTimeKind kind)
   at MySqlConnector.Core.Row.ParseDateTime(ReadOnlySpan`1 value)
   --- End of inner exception stack trace ---
   at MySqlConnector.Core.Row.ParseDateTime(ReadOnlySpan`1 value)
   at MySqlConnector.Core.Row.GetValue(Int32 ordinal)
   at MySql.Data.MySqlClient.MySqlDataReader.get_Item(String name)
   at TestNullDate.Program.Main(String[] args) in c:\\git\\TestNullDate\\TestNullDate\\Program.cs:line 29
```

**Comment by bgrainger** [comment]
Thanks for the code snippet; it's very helpful.

Connector/NET (MySql.Data) throws an `ArgumentOutOfRangeException` with this code (MySqlConnector throws a more appropriate IMO `FormatException`).

The workaround (for both libraries) is to add `AllowZeroDateTime=true` to your connection string. When you do that, all `DATE` and `DATETIME` columns are returned as `MySqlDateTime` objects, which can store "dates" that are illegal for `DateTime`. Unfortunately (in both libraries), the `IsValidDateTime` property only checks for `0000-00-00`; it doesn't validate that the month and day values are correct. You would have to add your own logic to check that.

I think this is working as designed; if you need to handle invalid dates, you should use the (somewhat misnamed) `AllowZeroDateTime=true` connection string option (which MySqlConnector keeps for backwards compatibility with Connector/NET).
